File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n
        forward = 1
        for i in range(n):
            res[i] *= forward
            forward *= nums[i]
        backwards = 1
        for i in range(n - 1, -1, -1):
            res[i] *= backwards
            backwards *= nums[i]
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
vector<vector<int>> res{};
    vector<vector<int>> combinationSum(const vector<int>& nums, int target) {
        vector<int> curr;
        backtrack(nums, target, curr, 0);
        return res;
    }

private:
    void backtrack(const vector<int>& nums, int target, vector<int>& curr, int i) {
        if (target == 0) {
            res.push_back(curr);
            return;
        }

        if (target < 0 || i >= nums.size()) return;

        curr.push_back(nums[i]);
        backtrack(nums, target - nums[i], curr, i);
        curr.pop_back();
        backtrack(nums, target, curr, i + 1);
    }
};
